Blog: why we ship faster than your company, probably (#6830)
PostHog/posthog.com: contents/founders/how-come-we-ship-so-much.md

---
date: 2023-10-30

> title: "Why we (probably) ship faster than your company"

Reviewer: ```suggestion
title: "We like to ship fast. Here's how we designed our company to do it"
```

author:
  - james-hawkins
featuredImage: ../images/blog/super-hog-pink.png
featuredImageType: full
tags:
  - Founders
  - Product engineers
  - Engineering
  - Culture
crosspost:
  - Product engineers
---

Clearly there are exceptions, but we ship a lot faster than the average company.

Our first line of code was January 22nd, 2020, after a pivot during a YC batch. Since then, we've shipped:

* [Product analytics](/product-analytics)
* [Session replay](/session-replay)
* [Feature flags](/feature-flags)
* [A/B testing](/ab-testing)
* [User surveys](/docs/surveys)

> * Heatmaps

Author: ```suggestion
* [Heatmaps](/docs/toolbar/heatmaps)
```

* [SDKs](/docs/libraries) for all the above
* A basic [CDP](/docs/cdp) to stream data to warehouses
* [Data warehouse](/docs/data-warehouse) MVP
* A wildly extensive website

...and those are just the things that worked.

All the above are being used a lot, too. We're ingesting 10s of billions of events a month, have 10s of thousands of companies regularly using our service, and over 100k user signups.

> I often get asked how we build so much, so quickly. The answer is simple: we deliberately designed our company for speed.

Author: ```suggestion
I often get asked how we build so much, so quickly. The answer is simple: we deliberately designed our company for speed.  We did this because we want to build [lots of products in one](/handbook/why-does-posthog-exist) - it's a lot to ship!
```

Today, we have people that can help engineers with design, but they are very happy to work reactively and fast – and they don't block merging work during QA. This means we do ship work where the design isn't perfect, but we won't drop things that are important and will keep improving them.

For example, we've redesigned this entire website ~5 times.

### 2. No product management by default

The responsibilities of a product manager still exist, but our engineers:

* Talk to users.
* Decide what to build within their product (and sometimes pitch whole new products).
* Have complete access to metrics like revenue, hosting costs, and (soon) staff costs, for their product.

The concept internally is a "[product engineer](/blog/what-is-a-product-engineer)" – i.e. an engineer who is product-minded and autonomous, rather than mindlessly filing tickets.

We have one product manager who steps in reactively if a small team needs more support.

> The above requires a lot of context setting from the company. To that end, we're *unusually transparent*. We share board slides, financial information, fundraising progress before, during and after it happens, and much more.

Reviewer: @jamesefhawkins Would you be up for adding some more detail here? Not essential, but nice-to-have. Alternatively, we can link to this: https://posthog.com/founders/how-to-run-a-transparent-company

> The above requires a lot of context setting from the company. To that end, we're *unusually transparent*. We share board slides, financial information, fundraising progress before, during and after it happens, and much more.

Author: ```suggestion
The above requires a lot of context setting from the company. To that end, we're *unusually transparent*. There are two main ways we do this.

First, we make clear what we're trying to achieve. To do this we have a clear, simple [mission and strategy](/handbook/why-does-posthog-exist). We communicate this repeatedly through our handbook, during everyone's onboarding, repeatedly in our all hands and when we plan each quarter of work. Our mission and strategy have both been refined as the company has made progress - they have felt more meaningful as a result as we've gotten further along. We've gradually gone from something hypothetical to these things feeling like an obvious label for what we do.

Second, when it comes to the _output_ of our work and our progress towards these goals, we share all our metrics (for each product and in total), financial information, fundraising progress (before, during and after it happens), and pretty much anything else people ask for.

We've written an entire post on [how to run a transparent company](/founders/how-to-run-a-transparent-company) that goes into more detail.
```

### 3. Hiring with intention

I think 90%+ of an organization's problems are solved by hiring right – see [what we've learned about hiring](https://newsletter.posthog.com/p/everything-weve-learned-about-hiring) in our newsletter.

We've found it invaluable to pay people during our hiring process to do some actual work. Over time, we've refined this down to a one-day task, which gives us a good sense of how much someone can get done and the quality of their actual code.

We tend to hire people with more experience, who we find work better with more autonomy. I think this is the magic combo when you're an all remote company, though we're working on ways to get better at growing people who are earlier in their careers.

### 4. Each team is a mini startup

I went to a superb talk by [Jeff Lawson](https://www.linkedin.com/in/jeffiel), who runs [Twilio](https://twilio.com/). He described how the greatest innovation at Amazon was that it felt like thousands of small startups. Simply speaking, a startup gets more done per person than a big corporate.

> PostHog works like a group of startups with lots of [small teams](/handbook/company/small-teams). Each of our products has a team of up to 6 who can ship with minimal interference from the rest of the company. We hire a lot of former technical founders to help here.

Reviewer: ```suggestion
PostHog works like a group of startups with lots of [small teams](/handbook/company/small-teams). Each of our products has a team of up to 6 who can ship with minimal interference from the rest of the company. They make their own product decisions, own their pricing, and are responsible for co-ordinating with our marketing and growth teams. We hire a lot of former technical founders to help here.
```

Reviewer: @jamesefhawkins Lior suggested we add some context around what being a "mini startup" means in practice. Does this work for you?

Author: yep

### 5. Nearly everyone is an engineer

It's simple, but if you have lots of spending and focus outside of engineering, you won't get as much built.

At the time of writing, [we are 37 people](/team), only 11 of whom are in Go To Market or Operations roles (and 6 of those 11 have at least some engineering in their background).

We don't do outbound sales, our [marketing team](/handbook/small-teams/marketing) is very small, we only have one product manager, and our [exec team](/handbook/small-teams/exec) is only 3 people. These things are all by design. 

We're product-led because our ideal users, engineers, want to try something rather than sit on a call.

### 6. No meeting Tuesdays and Thursdays

> We want a culture where people can deep work. We push back on unnecessary meetings. 

Reviewer: ```suggestion
We want a culture where people can deep work. We push back on unnecessary meetings and frequently review  existing meetings. 

If a meeting is booked for 45 minutes but frequently only takes 25, we'll rebook it so it's shorter by default – as was the case with out engineering sprint meetings recently.
```

We have learned, often painfully, that some meetings are necessary, but we default to asynchronous communication and adapt where needed.

> 

Reviewer 2: Another fun, related example is that the engineering sprint meeting was once 45 minutes but it usually took 25 minutes, so Tim just changed it to 25 minutes.

Reviewer: Added a line about this.

---
date: 2023-10-30

> title: "Why we (probably) ship faster than your company"

Reviewer: Yeah, discussed the same in Slack. Options I'm considering:

- How we designed our company for speed
- How to optimize your company for speed
- How and why we designed our company for speed
- We like to ship fast. So we designed our company around it.

> title: "Why we (probably) ship faster than your company"

Reviewer 3: I like the your 4th option the most – it makes me want to read it. The other 3 are quite dull and generic.

How about something along the lines of:
- We like to ship fast. Here's how we designed our company to do it

> title: "Why we (probably) ship faster than your company"

Reviewer: Noice. Works for me.

* [Product analytics](/product-analytics)
* [Session replay](/session-replay)
* [Feature flags](/feature-flags)
* [A/B testing](/ab-testing)
* [User surveys](/docs/surveys)
* Heatmaps
* [SDKs](/docs/libraries) for all the above
* A basic [CDP](/docs/cdp) to stream data to warehouses
* [Data warehouse](/docs/data-warehouse) MVP
* A wildly extensive website

...and those are just the things that worked.

All the above are being used a lot, too. We're ingesting 10s of billions of events a month, have 10s of thousands of companies regularly using our service, and over 100k user signups.

I often get asked how we build so much, so quickly. The answer is simple: we deliberately designed our company for speed.

## How we designed PostHog for speed

### 1. No design by default

When we started, we felt the best engineers would eventually quit if they felt dictated to by product designers. Product engineers need autonomy and freedom to enjoy their work.

As we grew, we realized that engineers can design the UX of a product. Especially if we hire people that have this skill set, and provide a framework for their work.

> We invested considerable time getting our design *system* up and running, and worked with a designer to get it done. Then we moved onto this "no design by default" phase. Had we been in this phase from the beginning, the app would have been very inconsistent... at best.

Reviewer: I agree, but probably not an important enough point to elaborate too much. Will look for a guide we can link to rather than adding more copy./

Today, we have people that can help engineers with design, but they are very happy to work reactively and fast – and they don't block merging work during QA. This means we do ship work where the design isn't perfect, but we won't drop things that are important and will keep improving them.

For example, we've redesigned this entire website ~5 times.

### 2. No product management by default

The responsibilities of a product manager still exist, but our engineers:

> * Talk to users.
> 
> * Decide what to build within their product (and sometimes pitch whole new products).
> 
> * Have complete access to metrics like revenue, hosting costs, and (soon) staff costs, for their product.

Reviewer: ```suggestion
* Talk to users.
* Decide what to build within their product (and sometimes pitch whole new products).
* Have complete access to metrics like revenue, hosting costs, and (soon) staff costs, for their product.
```

The concept internally is a "[product engineer](/blog/what-is-a-product-engineer)" – i.e. an engineer who is product-minded and autonomous, rather than mindlessly filing tickets.

We have one product manager who steps in reactively if a small team needs more support.

The above requires a lot of context setting from the company. To that end, we're *unusually transparent*. We share board slides, financial information, fundraising progress before, during and after it happens, and much more.

### 3. Hiring with intention

I think 90%+ of an organization's problems are solved by hiring right – see [what we've learned about hiring](https://newsletter.posthog.com/p/everything-weve-learned-about-hiring) in our Substack newsletter.

We've found it invaluable to pay people during our hiring process to do some actual work. Over time, we've refined this down to a one-day task, which gives us a good sense of how much someone can get done and the quality of their actual code.

> We tend to hire people with more experience, who we find work better with more autonomy. I think this is the magic combo when you're an all remote company, though we're working on ways to get better at growing people who are earlier in their careers.

Reviewer: I don't think so, tbh. They are different points.

### 4. Each team is a mini startup

I went to a superb talk by [Jeff Lawson](https://www.linkedin.com/in/jeffiel), who runs [Twilio](https://twilio.com/). He described how the greatest innovation at [Amazon](https://www.amazon.com/) was that it felt like thousands of small startups. Simply speaking, a startup gets more done per person than a big corporate.

> PostHog works like a group of startups with lots of [small teams](/handbook/company/small-teams). Each of our products has a team of up to 6 who can ship with minimal interference from the rest of the company. We hire a lot of former technical founders to help here.

Reviewer 3: This point doesn't tell me what we mean by a "mini startup". Can we add more detail? how do our teams operate? What makes it like a mini startup?

Reviewer: Good point. I can add this. Key points:

- They decide what to build
- They own their own pricing
- Responsible for monitoring churn, user feedback etc.
